sdk/dotnet/Pulumi.FSharp/Library.fs

```diff
@@ -1,15 +1,113 @@
-        |> Async.RunSynchronously
\ No newline at end of file
+        |> Async.RunSynchronously
+
+    /// <summary>
+    /// Runs an async function as a Pulumi <see cref="Deployment" />.
+    /// Blocks internally so that this function could be used directly from the main function.
+    /// </summary>
+    let runAsync (f: unit -> Async<IDictionary<string, obj>>) =
+        Deployment.RunAsync (fun () -> f() |> Async.StartAsTask)
+        |> Async.AwaitTask
+        |> Async.RunSynchronously
+
+
+/// <summary>
+/// Module containing utility functions to work with <see cref="Output{T}" />'s.
+/// </summary>
+module Outputs =
+    
+    /// <summary>
+    /// Transforms the data of <see cref="Output{'a}"/> with the provided function <paramref
+    /// name="f"/>. The result remains an <see cref="Output{'b}"/> so that dependent resources
+    /// can be properly tracked.
+    /// </summary>
+    let apply<'a, 'b> (f: 'a -> 'b) (output: Output<'a>) : Output<'b> =
+        output.Apply f
+
+    /// <summary>
+    /// Transforms the data of <see cref="Output{'a}"/> with the provided asynchronous function <paramref
+    /// name="f"/>. The result remains an <see cref="Output{'b}"/> so that dependent resources
+    /// can be properly tracked.
+    /// </summary>
+    let applyAsync<'a, 'b> (f: 'a -> Async<'b>) (output: Output<'a>) : Output<'b> =
+        output.Apply<'b> (f >> Async.StartAsTask)
+
+    /// <summary>
+    /// Transforms the data of <see cref="Output{'a}"/> with the provided function <paramref
+    /// name="f"/> that returns <see cref="Output{'b}"/>. The result is flattened to an <see cref="Output{'b}"/>.
+    /// </summary>
+    let bind<'a, 'b> (f: 'a -> Output<'b>) (output: Output<'a>) : Output<'b> =
+        output.Apply<'b> f
+    
+    /// <summary>
+    /// Combines an <see cref="Output{'a}"/> with an <see cref="Output{'b}"/> to produce an <see cref="Output{'a*'b}"/>.
+    /// </summary>
+    let pair<'a, 'b> (a: Output<'a>) (b: Output<'b>): Output<'a * 'b> =
+        Output.Tuple (a, b) 
+        |> apply (fun struct (a, b) -> (a, b))
+
+    /// <summary>
+    /// Combines three values of <see cref="Output"/> to produce a three-way tuple <see cref="Output"/>.
+    /// </summary>
+    let pair3<'a, 'b, 'c> (a: Output<'a>) (b: Output<'b>) (c: Output<'c>): Output<'a * 'b * 'c> =
+        Output.Tuple (io a, io b, io c) 
+        |> apply (fun struct (a, b, c) -> (a, b, c))
+
+    /// <summary>
+    /// Combines four values of <see cref="Output"/> to produce a four-way tuple <see cref="Output"/>.
+    /// </summary>
+    let pair4<'a, 'b, 'c, 'd> (a: Output<'a>) (b: Output<'b>) (c: Output<'c>) (d: Output<'d>): Output<'a * 'b * 'c * 'd> =
+        pair (pair a b) (pair c d)
+        |> apply(fun ((a, b), (c, d)) -> a, b, c, d)
+
+    /// <summary>
+    /// Combines a list of <see cref="Output{'a}"/> to produce an <see cref="Output{List{'a}}"/>.
+    /// </summary>
+    let all<'a> (values: List<Output<'a>>) : Output<List<'a>> = 
```

Comment:
Using (what appears to be) `system.collections.generic::List` here is reasonable, but forces the developer to use a particular _style_. Is there a strong reason not to make this more generic, and just take in a `seq<Output<'a>>` and return an `Output<seq<'a>>` instead?

It isn't the end of the world, but using the more primitive `seq<'a>` (`IEnumerable<T>`) gives the user more flexibility. Since then they could use `List`, `Seq`, `list (F#'s linked lists)`, an array, etc.

Reply:
It's a linked list (F# list) which is the default collection structure. C# uses arrays in `All`. 

`seq` in - `seq` out might suggest lazy evaluation of arguments, e.g. to be usable for unbound collections, so would be careful with this?

Reply:
Cool, if it's an F# list, then it's probably fine. (Since converting other data structures to/from list is common.)

I'd still argue that using a sequence would make things more efficient -- since the caller wouldn't need to convert, e.g. an array or `System.Collections.Generic::List` to an `FSharp.List` first. But I don't feel too strongly about it. You are right that `seq` also does imply lazy evaluation or the potential for being unbound... but if you are passing an unbound collection to `pulumi.all`, you kinda should expect it will lead to pain ;)

Anyways, sounds like it's fine to leave it as-is. 👍